stan/math/prim/mat/fun/ocl_gpu.hpp

```diff
@@ -0,0 +1,296 @@
+      oclContext_ = cl::Context(allDevices);
+      oclQueue_ = cl::CommandQueue(oclContext_, oclDevice_,
+                                   CL_QUEUE_PROFILING_ENABLE, NULL);
+      init_kernel_groups();
+      // Compile the dummy kernel used for timing purposes
+      cl::Program::Sources source(
+          1, std::make_pair(dummy_kernel.c_str(), dummy_kernel.size()));
+      cl::Program program_ = cl::Program(oclContext_, source);
+
+      try {
+        program_.build(allDevices);
+        kernels["dummy"] = cl::Kernel(program_, "dummy", NULL);
+        compiled_kernels["timing"] = true;
+      } catch (const cl::Error &e) {
+        domain_error(
+            "OpenCL Initialization", e.what(), e.err(),
+            "\nRetrieving build log\n",
+            program_.getBuildInfo<CL_PROGRAM_BUILD_LOG>(allDevices[0]).c_str());
+      }
+    } catch (const cl::Error &e) {
+      check_ocl_error("build", e);
+    }
+  }
+
+ public:
+  inline std::string description() const {
+    if (initialized_ == 1) {
+      return description_;
+    }
+    return "No device selected yet.";
+  }
+
+  inline cl::Context &context() {
+    if (initialized_ == 0) {
+      init();
+      initialized_ = 1;
+    }
+    return oclContext_;
+  }
+
+  inline cl::CommandQueue &queue() {
+    if (initialized_ == 0) {
+      init();
+      initialized_ = 1;
+    }
+    return oclQueue_;
+  }
+  inline int maxWorkgroupSize() { return max_workgroup_size; }
+};
+
+static ocl ocl_context_queue;
+
+/**
+ * Returns the description of the OpenCL
+ * platform and device that is used.
+ *
+ */
+inline std::string get_description() { return ocl_context_queue.description(); }
+
+/**
+ * Returns the reference to the
+ * OpenCL context. If no context was created,
+ * a new context is created.
+ *
+ */
+inline cl::Context &get_context() { return ocl_context_queue.context(); }
+/**
+ * Returns the reference to the active
+ * OpenCL command queue. If no context
+ * and queue were created,
+ * a new context and queue are created and
+ * the reference to the new queue is returned.
+ *
+ */
+inline cl::CommandQueue &get_queue() { return ocl_context_queue.queue(); }
+/**
+ * Returns the reference to the active
+ * OpenCL command queue. If no context
+ * and queue were created,
+ * a new context and queue are created and
+ * the reference to the new queue is returned.
+ *
+ */
+inline int get_maximum_workgroup_size() {
+  return ocl_context_queue.maxWorkgroupSize();
+}
+
+/**
+ * Compiles all the kernel in the specified group
+ *
+ * @param group The kernel group name
+ *
+ */
+inline void compile_kernel_group(std::string group) {
+  cl::Context &ctx = get_context();
+  std::vector<cl::Device> devices = ctx.getInfo<CL_CONTEXT_DEVICES>();
+  std::string kernel_source = kernel_strings[group];
+  cl::Program::Sources source(
+      1, std::make_pair(kernel_source.c_str(), kernel_source.size()));
+  cl::Program program_ = cl::Program(ctx, source);
+  try {
```

Comment:
can we move a bunch of this logic to one place so it's not executed every time?

Reply:
Srry what logic are you referring to? Also should I wait for you to review the new PR or should we go based on these comments?